src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -92,16 +102,81 @@ protected override Payload ConstructHeadersPayload(BlockLocator locator, uint256
             return headers;
         }
 
-        protected override void AttachCore()
-        {
-            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceivedAsync, true);
-        }
-
         /// <inheritdoc />
         public override object Clone()
         {
             return new ProvenHeadersConsensusManagerBehavior(
                 this.chain, this.initialBlockDownloadState, this.consensusManager, this.peerBanning, this.loggerFactory, this.network);
         }
+
+        /// <summary>
+        /// Determines whether the specified peer supports Proven Headers.
+        /// </summary>
+        /// <param name="peer">The peer.</param>
+        /// <returns>
+        ///   <c>true</c> if is peer is PH enabled; otherwise, <c>false</c>.
+        /// </returns>
+        private bool IsPeerPhEnabled(INetworkPeer peer)
+        {
+            return peer.Version >= NBitcoin.Protocol.ProtocolVersion.PROVEN_HEADER_VERSION;
+        }
+
+        /// <summary>
+        /// Determines whether the specified peer is Whitelisted.
+        /// </summary>
+        /// <param name="peer">The peer.</param>
+        /// <returns>
+        ///   <c>true</c> if the specified peer is Whitelisted; otherwise, <c>false</c>.
+        /// </returns>
+        private bool IsPeerWhitelisted(INetworkPeer peer)
+        {
+            return peer.Behavior<IConnectionManagerBehavior>()?.Whitelisted == true;
+        }
+
+        /// <summary>
+        /// Builds the GetHeadersPayload.
+        /// </summary>
+        /// <returns>The GetHeadersPayload instance.
+        /// If the peer can serve PH, GetProvenHeadersPayload is returned, otherwise if it's a legacy peer but it's whitelisted,
+        /// GetHeadersPayload is returned.
+        /// If the attached peer is a legacy peer and it's not whitelisted, returns null.
+        /// </returns>
+        protected override GetHeadersPayload BuildGetHeadersPayload()
+        {
+            INetworkPeer peer = this.AttachedPeer;
+
+            if (IsPeerPhEnabled(peer))
+            {
+                return new GetProvenHeadersPayload()
+                {
+                    BlockLocator = (this.ExpectedPeerTip ?? this.consensusManager.Tip).GetLocator(),
+                    HashStop = null
+                };
+            }
+            // If the peer doesn't supports PH but it's whitelisted, issue a standard GetHeadersPayload
```

Comment:
we should send normal headers to nodes that don't support them and PH to nodes that do. Doesn't matter if they are whitelisted or not.

Reply:
hmmm this raises another question when a legacy sends us headers we need to handle this (send them back headers) this will allow X to sync headers and blocks from # nodes

Reply:
when someone sends us legacy headers, `ProcessLegacyHeadersAsync` is executed and I think I should change
`if (!IsPeerPhEnabled(peer) && IsPeerWhitelisted(peer))`
to
`if (!IsPeerPhEnabled(peer))`